code/parse/sexp.cpp

```diff
@@ -2636,9 +2636,10 @@ int check_sexp_syntax(int node, int return_type, int recursive, int *bad_node, s
 						if (get_operator_const(op_node) < First_available_operator_id) {
 							ship_node = CDR(op_node);
 						} else {
-							ship_node = get_dynamic_parameter_index(Sexp_nodes[op_node].text, argnum);
+							ship_node = op_node + get_dynamic_parameter_index(Sexp_nodes[op_node].text, argnum); //error checking uses the whole tree
+							ship_node++; //dynamic params are saved at a 0 based index, but with the whole tree we need to move up a spot
 
-							if (ship_node < 0)
+							if (ship_node < op_node)
```

Comment:
This won't get the correct node in certain cases.  Sexp node iteration is based on Lisp, and in particular uses Lisp's CAR and CDR operations.  To get the *n*th argument, do `node = CDR(node)` *n* times.  The `(ship_node < 0)` check is correct because n will be negative if there are no subsequent nodes.

Reply:
That's actually not true here. I did a lot of testing and stepping through the code this morning to see exactly what was going on. The code I wrote provides the correct index from which to check the ship. And doing it with CDR sets `ship_node` to exactly the same integer as my original code. This is truly just you being anal retentive again. I look forward to the real world example where the code I wrote originally provides the incorrect node position.

Reply:
It is precisely the "real world" application that this guards against.  I wouldn't be surprised if, most of the time, the argument nodes are found in contiguous order.  But under heavy use the sexp node array could get fragmented, and in that situation you might find a gap in the node sequence, or a node appearing earlier in the Sexp_nodes array than its parent.  This is the nature of resource pools whose members are allocated and freed.  And if you ran across a bug in that real-world situation, it would be incredibly difficult to reproduce.  You yourself acknowledged that you got lucky in your original tests; this is just the same thing with another layer of subtlety.